conversation_reconstruction_dataflow/conversation_reconstruction/dataflow_main.py

```diff
@@ -48,83 +50,126 @@ def run(known_args, pipeline_args):
     '--staging_location=gs://wikidetox-viz-dataflow/staging',
     '--temp_location=gs://wikidetox-viz-dataflow/tmp',
     '--job_name=reconstruction-test',
-    '--num_workers=30'  
+    '--num_workers=5',
+    '--extra_package=third_party/mwparserfromhell.tar.gz'
   ])
-
-
   pipeline_options = PipelineOptions(pipeline_args)
   pipeline_options.view_as(SetupOptions).save_main_session = True
+  renaming = "latest.rev_id_in_int as rev_id_in_int, latest.week as week, latest.year as year, latest.sha1 as sha1, latest.user_id as user_id, latest.format as format, latest.user_text as user_text, latest.timestamp as timestamp, latest.text as text, latest.page_title as page_title, latest.model as model, latest.page_namespace as page_namespace, latest.page_id as page_id, latest.rev_id as rev_id, latest.comment as comment, latest.user_ip as user_ip, latest.truncated as truncated, latest.records_count as records_count, latest.record_index as record_index"
+  get_max_query= "SELECT * FROM (SELECT * FROM %s, (SELECT %s FROM (SELECT %s FROM %s as latest INNER JOIN (SELECT max(rev_id_in_int) as max_rev_id FROM %s WHERE week < %d or year < %d GROUP BY week, year, page_id) as previous_max"%(known_args.input_table, renaming, renaming, known_args.input_table, known_args.input_table, known_args.week, known_args.year)
```

Comment:
OK, for these string substitutions, definitely used name params, otherwise it's way to complex to read (code is read more often than written, so saving reading time is more important than writing time) .  :)

Reply:
Yep, really good advice should have taken this earlier today, had a really painful time debugging this.